### PR TITLE
Alerting: Handle NoData and Error evaluation results

### PR DESCRIPTION
**Pull request by davidmparrott**
**What this PR does / why we need it**:

- Adds support for evaluation results that have NoData or Error set as their state
- Refactor setNextState to be more readable
- Rework unit tests
- Add integration test for /rules from @gotjosh 

**Which issue(s) this PR fixes**:
https://github.com/grafana/alerting-squad/issues/56

**Special notes for your reviewer**:
Addresses some feedback from https://github.com/grafana/grafana/pull/33085

### CODE REVIEW

File: pkg/services/ngalert/api/api_prometheus.go [hunk 1]
```diff
@@ -106,36 +98,45 @@ func (srv PrometheusSrv) RouteGetRuleStatuses(c *models.ReqContext) response.Res
...
+				case eval.Pending:
 					if alertingRule.State == "inactive" {
-						alertingRule.State = "pending"
+						alertingRule.State = "firing"
 					}
-				case "firing":
+				case eval.Alerting:
 					alertingRule.State = "firing"
+				case eval.Error:
+					// handle Error case based on configuration in alertRule
```
Comment (kylebrandt) [inline pkg/services/ngalert/api/api_prometheus.go:133]: This and the following TODOs?

File: pkg/services/ngalert/eval/eval.go [hunk 2]
```diff
@@ -76,6 +76,11 @@ const (
 	// that evaluated to true (Alerting).
 	Alerting
 
+	// Pending is the eval state for an alert instance condition
+	// that evaluated to true (Alerting) but has not yet met
+	// the For duration defined in AlertRule
```
Comment (kylebrandt) [inline pkg/services/ngalert/eval/eval.go:81]:
```suggestion
	// the For duration defined in AlertRule.
```

File: pkg/services/ngalert/api/api_prometheus.go (hunk 1 above)
Comment (davidmparrott) [reply pkg/services/ngalert/api/api_prometheus.go]: I'm going to change the health of the alertingRule but have to think about/discuss if there's anything else we should do in these cases.

File: pkg/services/ngalert/state/state_tracker.go [hunk 3]
```diff
@@ -136,57 +134,120 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
...
+			a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+			return a
+		}
+		a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+	case eval.Pending:
+		if result.EvaluatedAt.Sub(a.StartsAt) > alertRule.For {
+			a.State = eval.Alerting
+			a.StartsAt = result.EvaluatedAt
+			a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+			a.Annotations["alerting_at"] = result.EvaluatedAt.String()
```
Comment (gotjosh) [inline pkg/services/ngalert/state/state_tracker.go:186]: What's the reasoning behind this custom annonnation?

File: pkg/tests/api/alerting/api_prometheus_test.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+			err := resp.Body.Close()
+			require.NoError(t, err)
+		})
+		b, err := ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		assert.Equal(t, 200, resp.StatusCode)
+		require.JSONEq(t, `{"status": "success", "data": {"groups": []}}`, string(b))
+	}
+
+	// Now, let's create some rules
```
Comment (gotjosh) [inline pkg/tests/api/alerting/api_prometheus_test.go:51]: Is there any chance, we could add two example rules that would result in no data or error to see how they'd look like?

File: pkg/tests/api/alerting/api_prometheus_test.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+									"datasourceUid": "-100",
+									"type": "math",
+									"expression": "2 + 3 > 1"
+									}`),
+							},
+						},
+						NoDataState:  apimodels.NoDataState(ngmodels.Alerting),
+						ExecErrState: apimodels.ExecutionErrorState(ngmodels.KeepLastStateErrState),
+					},
+				},
```
Comment (gotjosh) [inline pkg/tests/api/alerting/api_prometheus_test.go:104]:
```suggestion
				{
					GrafanaManagedAlert: &apimodels.PostableGrafanaRule{
						Title:     "NeverFiring",
						Condition: "A",
						Data: []ngmodels.AlertQuery{
							{
								RefID: "A",
								RelativeTimeRange: ngmodels.RelativeTimeRange{
									From: ngmodels.Duration(time.Duration(5) * time.Hour),
									To:   ngmodels.Duration(time.Duration(3) * time.Hour),
								},
								Model: json.RawMessage(`{
									"datasourceUid": "-100",
									"type": "math",
									"expression": "2 + 3 > 10"
									}`),
							},
						},
						NoDataState:  apimodels.NoDataState(ngmodels.Alerting),
						ExecErrState: apimodels.ExecutionErrorState(ngmodels.KeepLastStateErrState),
					},
				},
```

File: pkg/services/ngalert/state/state_tracker.go [hunk 6]
```diff
@@ -136,57 +134,120 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
...
+		a.StartsAt = result.EvaluatedAt
+		if !(alertRule.For > 0) {
+			a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+			a.State = eval.Alerting
+			a.Annotations["alerting_at"] = result.EvaluatedAt.String()
+		} else {
+			a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+			if result.EvaluatedAt.Sub(a.StartsAt) > alertRule.For {
+				a.State = eval.Alerting
+				a.Annotations["alerting_at"] = result.EvaluatedAt.String()
```
Comment (gotjosh) [inline pkg/services/ngalert/state/state_tracker.go:198]: Ditto

File: pkg/services/ngalert/state/state_tracker.go [hunk 7]
```diff
@@ -136,57 +134,120 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
...
+	if a.StartsAt.IsZero() {
+		a.StartsAt = result.EvaluatedAt
+	}
+	if !(alertRule.For > 0) {
+		a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+	} else {
+		a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+	}
+	if a.State != eval.Error {
+		a.Annotations["last_error"] = result.EvaluatedAt.String()
```
Comment (gotjosh) [inline pkg/services/ngalert/state/state_tracker.go:217]: Is this correct? I'm just curious about whenever we meant to sent last_error as timestamp here.

I'm a bit puzzled by the inclusion of these annotations - they might be correct, I just want to get an understanding of why.

File: pkg/services/ngalert/state/state_tracker.go [hunk 8]
```diff
@@ -136,57 +134,120 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
...
+	if a.StartsAt.IsZero() {
+		a.StartsAt = result.EvaluatedAt
+	}
+	if !(alertRule.For > 0) {
+		a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+	} else {
+		a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+	}
+	if a.State != eval.NoData {
+		a.Annotations["no_data"] = result.EvaluatedAt.String()
```
Comment (gotjosh) [inline pkg/services/ngalert/state/state_tracker.go:238]: Similar comment here.

**Review by gotjosh — COMMENTED** [review]
Thanks David for taking into consideration some of my changes here 🎉 

Made a first pass on the state tracker. I haven't looked at the integration test but will do once I have a better understanding of what's going on.

**Review by gotjosh — APPROVED** [review]
I'm going to unblock this work now that I have understood it a bit better. ✨

For others, the main thing this is trying to achieve is handling the case where an evaluation result produces no data or error. On both scenarios (but depending on configuration) this should transition the evaluation result to "alerting" thus sending an alert to the notification system.

I believe this does not cover:

- How this would look like at a Prometheus-compatible API level
- Any information that could give us an idea on why this is alerting when the notification is sent
- An integration test that could cover the various cases on which this would trigger a notification and how that would look like by querying the /alerts endpoint
- The minimum re-send delay implementation or correctly setting the `EndsAt` on all known cases

I believe these have been discussed and will follow up on subsequent work.

**Comment by codesome** [comment]
> On both scenarios (but depending on configuration) this should transition the evaluation result to "alerting" thus sending an alert to the notification system.

Regarding no-data, IIUC, it is an explicit option in rules to alert on no-data and not done for all the rules. Because by default we go to pending and alerting when we _see some data_. Am I understanding this right?

**Comment by davidmparrott** [comment]
@codesome that sounds correct. I'll elaborate just to make sure we're on the same page. 

Evaluation results will have one of four states: alerting, normal, nodata, error https://github.com/grafana/grafana/blob/master/pkg/services/ngalert/eval/eval.go#L222-L230

Those evaluation results are passed to the state manager and the state of the cache entry is updated or not based on the state in the evaluation result, the state of the cache entry, and the configuration of the alert rule. https://github.com/grafana/grafana/blob/master/pkg/services/ngalert/state/state.go#L30-L119

For an evaluation result of normal, if the state is anything other than normal, the state is set to normal and EndsAt is updated to the time that the evaluation result was created.

For an evaluation result of alerting:

- If the cache entry is already alerting, the EndsAt time is extended
- If the entry is pending, and we have exceeded the For set in the alert rule, the entry is changed to alerting, EndsAt is updated, and StartsAt is set to the time the evaluation result was created
- For any other state (normal, error, nodata) the cache entry is set to alerting, ends at and starts at are updated (depending on whether For is set or not)

For an evaluation result of nodata:

- If no starts at is set, starts at is set to the time the evaluation result was created
- If for is set, endsat is set to be the duraiton of For
- If For is not set, endsat is set to 2X the interval duration
- Depending on the NoDataState set in the alert rule, the state is set to alerting, nodata, normal, or the state is not changed in the event that KeepLastState is set. Note: this is where we run into edge cases mentioned in the docv @kylebrandt created before we started this project.

For an evaluation result of error:

- Logic is very similar to NoData except there are fewer possible options for AlertingErrState. We can keep last state or set to alerting.

**Pull request merged**